lib/index.ts

```diff
@@ -0,0 +1,413 @@
+import $Refs from "./refs.js";
+import _parse from "./parse.js";
+import normalizeArgs from "./normalize-args.js";
+import resolveExternal from "./resolve-external.js";
+import _bundle from "./bundle.js";
+import _dereference from "./dereference.js";
+import * as url from "./util/url.js";
+import {
+  JSONParserError,
+  InvalidPointerError,
+  MissingPointerError,
+  ResolverError,
+  ParserError,
+  UnmatchedParserError,
+  UnmatchedResolverError,
+  isHandledError,
+  JSONParserErrorGroup,
+} from "./util/errors.js";
+import { ono } from "@jsdevtools/ono";
+import maybe from "./util/maybe.js";
+import type { ParserOptions } from "./options.js";
+import type { $RefsCallback, JSONSchema, SchemaCallback } from "./types/index.js";
+
+export { JSONParserError };
+export { InvalidPointerError };
+export { MissingPointerError };
+export { ResolverError };
+export { ParserError };
+export { UnmatchedParserError };
+export { UnmatchedResolverError };
+
+type RefParserSchema = string | JSONSchema;
+
+/**
+ * This class parses a JSON schema, builds a map of its JSON references and their resolved values,
+ * and provides methods for traversing, manipulating, and dereferencing those references.
+ *
+ * @class
+ */
+export class $RefParser {
+  /**
+   * The parsed (and possibly dereferenced) JSON schema object
+   *
+   * @type {object}
+   * @readonly
+   */
+  public schema: JSONSchema | null = null;
+
+  /**
+   * The resolved JSON references
+   *
+   * @type {$Refs}
+   * @readonly
+   */
+  $refs = new $Refs();
+
+  /**
+   * Parses the given JSON schema.
+   * This method does not resolve any JSON references.
+   * It just reads a single file in JSON or YAML format, and parse it as a JavaScript object.
+   *
+   * @param [path] - The file path or URL of the JSON schema
+   * @param [schema] - A JSON schema object. This object will be used instead of reading from `path`.
+   * @param [options] - Options that determine how the schema is parsed
+   * @param [callback] - An error-first callback. The second parameter is the parsed JSON schema object.
+   * @returns - The returned promise resolves with the parsed JSON schema object.
+   */
+  public parse(schema: RefParserSchema): Promise<JSONSchema>;
+  public parse(schema: RefParserSchema, callback: SchemaCallback): Promise<void>;
```

Comment:
I've not done method overloading in TypeScript before. Do they all have to have the same return value, or should the variants with a callback return void instead of Promise<void>?

Reply:
Well the types that are currently used in the repo are actually incorrect - all the function calls return a Promise, since it's an async function. This just types it correctly.

However, with this type-definition overloading, it's saying that when you pass in just `schema`, the promise will resolve to a `JSONSchema`, whereas if you pass in a callback the callback will be called and the promise will be void.

I don't really like the usage of `arguments` in all these functions, and it would be nice to be able to write new APIs that actually do satisfy some of these apis. They're written in a very "accepting" way right now, IMO.